cartography/intel/okta/applications.py

```diff
@@ -181,34 +180,59 @@ def transform_okta_application_list(okta_applications):
 
 
 def transform_okta_application(okta_application):
-    # https://github.com/okta/okta-sdk-python/blob/master/okta/models/app/AppInstance.py
     app_props = {}
-    app_props["id"] = okta_application.id
-    app_props["name"] = okta_application.name
-    app_props["label"] = okta_application.label
-    if okta_application.created:
-        app_props["created"] = okta_application.created.strftime("%m/%d/%Y, %H:%M:%S")
+    app_props["id"] = okta_application["id"]
+    app_props["name"] = okta_application["name"]
+    app_props["label"] = okta_application["label"]
+    if "created" in okta_application and okta_application["created"]:
+        app_props["created"] = datetime.strptime(
+            okta_application["created"], "%Y-%m-%dT%H:%M:%S.%fZ",
+        ).strftime("%m/%d/%Y, %H:%M:%S")
     else:
         app_props["created"] = None
 
-    if okta_application.lastUpdated:
-        app_props["okta_last_updated"] = okta_application.lastUpdated.strftime("%m/%d/%Y, %H:%M:%S")
+    if "lastUpdated" in okta_application and okta_application["lastUpdated"]:
+        app_props["okta_last_updated"] = datetime.strptime(
+            okta_application["lastUpdated"], "%Y-%m-%dT%H:%M:%S.%fZ",
+        ).strftime("%m/%d/%Y, %H:%M:%S")
     else:
         app_props["okta_last_updated"] = None
 
-    app_props["status"] = okta_application.status
+    app_props["status"] = okta_application["status"]
 
-    if okta_application.activated:
-        app_props["activated"] = okta_application.activated.strftime("%m/%d/%Y, %H:%M:%S")
+    if "activated" in okta_application and okta_application["activated"]:
+        app_props["activated"] = datetime.strptime(
+            okta_application["activated"], "%Y-%m-%dT%H:%M:%S.%fZ",
+        ).strftime("%m/%d/%Y, %H:%M:%S")
     else:
         app_props["activated"] = None
 
-    app_props["features"] = okta_application.features
-    app_props["sign_on_mode"] = okta_application.signOnMode
+    app_props["features"] = okta_application["features"]
+    app_props["sign_on_mode"] = okta_application["signOnMode"]
 
     return app_props
 
 
+def transform_okta_application_extract_replyurls(okta_application):
+    """
+    Extracts the reply uri information from an okta app
+    and determines if the dns of the reply url is valid
+    """
+    uris = []
+    if "oauthClient" in okta_application["settings"]:
+        if "redirect_uris" in okta_application["settings"]["oauthClient"]:
+            for uri in okta_application["settings"]["oauthClient"]["redirect_uris"]:
+                netloc = urlparse(uri).netloc
+                try:
+                    socket.gethostbyname(netloc)
```

Comment:
Ooh interesting...

We talked about this a bit in person.  I really like that we're starting to explore more "active" forms of data validation.  At bare minimum for safety and speed reasons, I think this step needs a toggle switch with its default set to "do not perform gethostbyname()".

I almost think this kind of functionality might fit in better with a separate tool that would integrate with Cartography: kind of an "active reconnaissance data validator".

CC: @ecdavis

Reply:
I would really like to get this in, so if I need to pull this functionality I can. Personally i think it fits as its passive data gathering. We could add in a cli parameter and opt-in or opt-out of these queries.